src/eval/operation.rs

```diff
@@ -1850,17 +1854,16 @@ fn process_binary_operation(
                             debug_assert!(attrs1.closurized, "the left-hand side of ArrayConcat (@) is not closurized.");
                             debug_assert!(attrs2.closurized, "the right-hand side of ArrayConcat (@) is not closurized.");
 
-                            let mut ts: Vec<RichTerm> = Vec::with_capacity(ts1.len() + ts2.len());
-
-                            ts.extend(ts1.into_iter());
-                            ts.extend(ts2.into_iter());
+                            // TODO: chain tends to be slow, but Array currently can only be constructed from iterators.
+                            // it's possible to use Rc<[T]>::new_uninit_slice() (nightly-only) and fill up the array manually.
+                            let ts = ts1.into_iter().chain(ts2.into_iter()).collect();
```

Comment:
I wonder if one can get an owned slice out of a Vector (even it that implies `mem::swap` or the like). I couldn't find something with a quick research though. If it were to be true, we could use an array as a buffer and then just get back the slice to build a Nickel array.

Reply:
It looks like because an `Array` iterator does not implement `TrustedLen` (an experimental nightly-only API), collecting into an `Rc<[RichTerm]>` will go through a vector anyway. So might as well use a vector as a buffer and then use `Rc::from` like the standard library does.

https://doc.rust-lang.org/std/rc/struct.Rc.html#impl-FromIterator%3CT%3E